core/cluster/process.go

```diff
@@ -0,0 +1,38 @@
+package cluster
+
+import "time"
+
+type (
+
+	SingleProcess interface {
+		//Do runs the given function if there is no other function with that key running
+		Do(key string, fn func()) (bool, error)
+	}
+
+	DistributedLock interface {
+		//GetLock - call to requests a distributed lock - if not possible. returns "false" in case Lock is not available without blocking
+		GetLock(key string, max *time.Duration) (bool, error)
+		//Free the distributed Lock
+		Unlock(key string)
+	}
+
+	SingleProcessImpl struct {
+		lock DistributedLock
+	}
+)
+
+//Do runs the func if Lock can be obtained
+func (s *SingleProcessImpl) Do(key string, fn func()) (bool, error) {
```

Comment:
Unclear what a result of `false, nil` means: not run but no error? Then `Do` would be the wrong naming, as it might or might not run the function.